File: apps/gateway/src/db/agent.ts
import { and, asc, desc, eq, gt, max } from 'drizzle-orm';
import { getDb as getOrmDb } from './client';
import {
  type AgentConfirmationStatus,
  type AgentMessageRole,
  type AgentSessionStatus,
  type AgentWriteMode,
  agentConfirmations,
  agentMessages,
  agentSessions,
  agentSettings,
} from './schema';

export type AgentSettingsRecord = typeof agentSettings.$inferSelect;
export type AgentSessionRecord = typeof agentSessions.$inferSelect;
export type AgentMessageRecord = typeof agentMessages.$inferSelect;
export type AgentConfirmationRecord = typeof agentConfirmations.$inferSelect;

export type {
  AgentConfirmationStatus,
  AgentMessageRole,
  AgentSearchProvider,
  AgentSessionStatus,
  AgentWriteMode,
} from './schema';

export function ensureAgentSettingsInitialized(): void {
  const orm = getOrmDb();

  orm
    .insert(agentSettings)
    .values({
      id: 1,
      searchProvider: 'none',
      tavilyApiKeyEnc: null,
      braveApiKeyEnc: null,
      defaultProviderId: null,
      defaultModelId: null,
      updatedAt: new Date().toISOString(),
    })
    .onConflictDoNothing({ target: agentSettings.id })
    .run();
}

export function getAgentSettings(): AgentSettingsRecord {
  const orm = getOrmDb();
  let row = orm.select().from(agentSettings).where(eq(agentSettings.id, 1)).get();

  if (!row) {
    ensureAgentSettingsInitialized();
    row = orm.select().from(agentSettings).where(eq(agentSettings.id, 1)).get();
  }

  if (!row) {
    throw new Error('agent_settings not initialized');
  }

  return row;
}

export function updateAgentSettings(
  updates: Partial<Omit<AgentSettingsRecord, 'id' | 'updatedAt'>>
): AgentSettingsRecord {
  getAgentSettings();

  const orm = getOrmDb();
  const setValues: Partial<typeof agentSettings.$inferInsert> = {
    updatedAt: new Date().toISOString(),
  };

  if (updates.searchProvider !== undefined) {
    setValues.searchProvider = updates.searchProvider;
  }
  if (updates.tavilyApiKeyEnc !== undefined) {
    setValues.tavilyApiKeyEnc = updates.tavilyApiKeyEnc;
  }
  if (updates.braveApiKeyEnc !== undefined) {
    setValues.braveApiKeyEnc = updates.braveApiKeyEnc;
  }
  if (updates.defaultProviderId !== undefined) {
    setValues.defaultProviderId = updates.defaultProviderId;
  }
  if (updates.defaultModelId !== undefined) {
    setValues.defaultModelId = updates.defaultModelId;
  }

  orm.update(agentSettings).set(setValues).where(eq(agentSettings.id, 1)).run();
  return getAgentSettings();
}

export interface CreateAgentSessionInput {
  title: string;
  deviceId?: string | null;
  paneId?: string | null;
  providerId?: string | null;
  modelId: string;
  systemPrompt?: string | null;
  writeMode?: AgentWriteMode;
  useProviderWebSearch?: boolean;
  maxStepsPerTurn?: number;
}

export function createAgentSession(input: CreateAgentSessionInput): AgentSessionRecord {
  const orm = getOrmDb();
  const now = new Date().toISOString();
  const row: typeof agentSessions.$inferInsert = {
    id: crypto.randomUUID(),
    title: input.title,
    deviceId: input.deviceId ?? null,
    paneId: input.paneId ?? null,
    providerId: input.providerId ?? null,
    modelId: input.modelId,
    systemPrompt: input.systemPrompt ?? null,
    writeMode: input.writeMode ?? 'confirm',
    useProviderWebSearch: input.useProviderWebSearch ?? false,
    status: 'idle',
    lastError: null,
    maxStepsPerTurn: input.maxStepsPerTurn ?? 25,
    createdAt: now,
    updatedAt: now,
  };

  orm.insert(agentSessions).values(row).run();
  const created = getAgentSessionById(row.id);
  if (!created) {
    throw new Error('failed to create agent session');
  }
  return created;
}

export function getAgentSessionById(id: string): AgentSessionRecord | null {
  const orm = getOrmDb();
  return orm.select().from(agentSessions).where(eq(agentSessions.id, id)).get() ?? null;
}

export function getAllAgentSessions(): AgentSessionRecord[] {
  const orm = getOrmDb();
  return orm.select().from(agentSessions).orderBy(desc(agentSessions.updatedAt)).all();
}

export function getAgentSessionsByStatus(status: AgentSessionStatus): AgentSessionRecord[] {
  const orm = getOrmDb();
  return orm
    .select()
    .from(agentSessions)
    .where(eq(agentSessions.status, status))
    .orderBy(desc(agentSessions.updatedAt))
    .all();
}

export function updateAgentSession(
  id: string,
  updates: Partial<Omit<AgentSessionRecord, 'id' | 'createdAt' | 'updatedAt'>>
): AgentSessionRecord | null {
  const orm = getOrmDb();
  const setValues: Partial<typeof agentSessions.$inferInsert> = {
    updatedAt: new Date().toISOString(),
  };

  for (const key of [
    'title',
    'deviceId',
    'paneId',
    'providerId',
    'modelId',
    'systemPrompt',
    'writeMode',
    'useProviderWebSearch',
    'status',
    'lastError',
    'maxStepsPerTurn',
  ] as const) {
    if (updates[key] !== undefined) {
      (setValues as Record<string, unknown>)[key] = updates[key];
    }
  }

  orm.update(agentSessions).set(setValues).where(eq(agentSessions.id, id)).run();
  return getAgentSessionById(id);
}

export function deleteAgentSession(id: string): void {
  const orm = getOrmDb();
  orm.delete(agentSessions).where(eq(agentSessions.id, id)).run();
}

export function appendAgentMessage(
  sessionId: string,
  role: AgentMessageRole,
  content: unknown
): AgentMessageRecord {
  const orm = getOrmDb();
  const id = crypto.randomUUID();

  orm.transaction((tx) => {
    const current = tx
      .select({ maxSeq: max(agentMessages.seq) })
      .from(agentMessages)
      .where(eq(agentMessages.sessionId, sessionId))
      .get();

    tx.insert(agentMessages)
      .values({
        id,
        sessionId,
        seq: (current?.maxSeq ?? -1) + 1,
        role,
        content,
        createdAt: new Date().toISOString(),
      })
      .run();
  });

  const created = orm.select().from(agentMessages).where(eq(agentMessages.id, id)).get();
  if (!created) {
    throw new Error('failed to append agent message');
  }
  return created;
}

export function listAgentMessages(
  sessionId: string,
  options: { afterSeq?: number } = {}
): AgentMessageRecord[] {
  const orm = getOrmDb();
  const conditions =
    options.afterSeq !== undefined
      ? and(eq(agentMessages.sessionId, sessionId), gt(agentMessages.seq, options.afterSeq))
      : eq(agentMessages.sessionId, sessionId);

  return orm.select().from(agentMessages).where(conditions).orderBy(asc(agentMessages.seq)).all();
}

export function getMaxAgentMessageSeq(sessionId: string): number {
  const orm = getOrmDb();
  const row = orm
    .select({ maxSeq: max(agentMessages.seq) })
    .from(agentMessages)
    .where(eq(agentMessages.sessionId, sessionId))
    .get();
  return row?.maxSeq ?? -1;
}

export interface CreateAgentConfirmationInput {
  sessionId: string;
  toolName: string;
  toolCallId: string;
  inputJson: unknown;
}

export function createAgentConfirmation(
  input: CreateAgentConfirmationInput
): AgentConfirmationRecord {
  const orm = getOrmDb();
  const row: typeof agentConfirmations.$inferInsert = {
    id: crypto.randomUUID(),
    sessionId: input.sessionId,
    toolName: input.toolName,
    toolCallId: input.toolCallId,
    inputJson: input.inputJson,
    status: 'pending',
    reason: null,
    decidedAt: null,
    createdAt: new Date().toISOString(),
  };

  orm.insert(agentConfirmations).values(row).run();
  const created = getAgentConfirmationById(row.id);
  if (!created) {
    throw new Error('failed to create agent confirmation');
  }
  return created;
}

export function getAgentConfirmationById(id: string): AgentConfirmationRecord | null {
  const orm = getOrmDb();
  return orm.select().from(agentConfirmations).where(eq(agentConfirmations.id, id)).get() ?? null;
}

export function listPendingAgentConfirmations(sessionId: string): AgentConfirmationRecord[] {
  const orm = getOrmDb();
  return orm
    .select()
    .from(agentConfirmations)
    .where(
      and(eq(agentConfirmations.sessionId, sessionId), eq(agentConfirmations.status, 'pending'))
    )
    .orderBy(asc(agentConfirmations.createdAt))
    .all();
}

export function decideAgentConfirmation(
  id: string,
  decision: { status: Exclude<AgentConfirmationStatus, 'pending'>; reason?: string | null }
): AgentConfirmationRecord | null {
  const orm = getOrmDb();
  const updated = orm
    .update(agentConfirmations)
    .set({
      status: decision.status,
      reason: decision.reason ?? null,
      decidedAt: new Date().toISOString(),
    })
    .where(and(eq(agentConfirmations.id, id), eq(agentConfirmations.status, 'pending')))
    .returning()
    .get();

  return updated ?? null;
}

File: apps/gateway/src-rust/database/repository/agent.rs
use sea_orm::entity::prelude::Json;
use sea_orm::sea_query::OnConflict;
use sea_orm::{ColumnTrait, EntityTrait, QueryFilter, QueryOrder, QuerySelect, Set};
use serde_json::json;

use crate::entity::{
    agent_confirmations, agent_messages, agent_queued_messages, agent_sessions, agent_settings,
};

use super::{
    bool_value, json_string_list, new_id, now_iso, rollback, Repository, RepositoryError,
    RepositoryResult,
};

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct AgentSettingsUpdate {
    pub search_provider: Option<String>,
    pub tavily_api_key_enc: Option<Option<String>>,
    pub brave_api_key_enc: Option<Option<String>>,
    pub default_provider_id: Option<Option<String>>,
    pub default_model_id: Option<Option<String>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CreateAgentSessionInput {
    pub title: String,
    pub device_id: Option<String>,
    pub pane_id: Option<String>,
    pub provider_id: Option<String>,
    pub model_id: String,
    pub system_prompt: Option<String>,
    pub write_mode: Option<String>,
    pub use_provider_web_search: Option<bool>,
    pub provider_hosted_tools: Option<Vec<String>>,
    pub allow_control_chars: Option<bool>,
    pub origin_pane_title: Option<String>,
    pub origin_process_name: Option<String>,
    pub max_steps_per_turn: Option<i64>,
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct AgentSessionUpdate {
    pub title: Option<String>,
    pub device_id: Option<Option<String>>,
    pub pane_id: Option<Option<String>>,
    pub provider_id: Option<Option<String>>,
    pub model_id: Option<String>,
    pub system_prompt: Option<Option<String>>,
    pub write_mode: Option<String>,
    pub use_provider_web_search: Option<bool>,
    pub provider_hosted_tools: Option<Vec<String>>,
    pub allow_control_chars: Option<bool>,
    pub status: Option<String>,
    pub last_error: Option<Option<String>>,
    pub max_steps_per_turn: Option<i64>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct CreateAgentConfirmationInput {
    pub id: Option<String>,
    pub session_id: String,
    pub tool_name: String,
    pub tool_call_id: String,
    pub input_json: Json,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct AgentConfirmationDecision {
    pub status: String,
    pub reason: Option<String>,
}

impl Repository {
    pub async fn ensure_agent_settings_initialized(&self) -> RepositoryResult<()> {
        agent_settings::Entity::insert(agent_settings::ActiveModel {
            id: Set(1),
            search_provider: Set("none".to_owned()),
            tavily_api_key_enc: Set(None),
            brave_api_key_enc: Set(None),
            default_provider_id: Set(None),
            default_model_id: Set(None),
            updated_at: Set(now_iso()),
        })
        .on_conflict(
            OnConflict::column(agent_settings::Column::Id)
                .do_nothing()
                .to_owned(),
        )
        .exec_without_returning(self.database.orm())
        .await?;
        Ok(())
    }

    pub async fn get_agent_settings(&self) -> RepositoryResult<agent_settings::Model> {
        if let Some(model) = agent_settings::Entity::find_by_id(1)
            .one(self.database.orm())
            .await?
        {
            return Ok(model);
        }
        self.ensure_agent_settings_initialized().await?;
        agent_settings::Entity::find_by_id(1)
            .one(self.database.orm())
            .await?
            .ok_or(RepositoryError::NotInitialized("agent_settings"))
    }

    pub async fn update_agent_settings(
        &self,
        updates: AgentSettingsUpdate,
    ) -> RepositoryResult<agent_settings::Model> {
        self.get_agent_settings().await?;
        let mut model = agent_settings::ActiveModel {
            id: Set(1),
            updated_at: Set(now_iso()),
            ..Default::default()
        };
        if let Some(value) = updates.search_provider {
            model.search_provider = Set(value);
        }
        if let Some(value) = updates.tavily_api_key_enc {
            model.tavily_api_key_enc = Set(value);
        }
        if let Some(value) = updates.brave_api_key_enc {
            model.brave_api_key_enc = Set(value);
        }
        if let Some(value) = updates.default_provider_id {
            model.default_provider_id = Set(value);
        }
        if let Some(value) = updates.default_model_id {
            model.default_model_id = Set(value);
        }
        agent_settings::Entity::update(model)
            .exec(self.database.orm())
            .await?;
        self.get_agent_settings().await
    }

    pub async fn create_agent_session(
        &self,
        input: CreateAgentSessionInput,
    ) -> RepositoryResult<agent_sessions::Model> {
        let now = now_iso();
        let model = agent_sessions::Model {
            id: new_id(),
            title: input.title,
            device_id: input.device_id,
            pane_id: input.pane_id,
            provider_id: input.provider_id,
            model_id: input.model_id,
            system_prompt: input.system_prompt,
            write_mode: input.write_mode.unwrap_or_else(|| "confirm".to_owned()),
            use_provider_web_search: bool_value(input.use_provider_web_search.unwrap_or(false)),
            provider_hosted_tools: json_string_list(
                &input.provider_hosted_tools.unwrap_or_default(),
            ),
            allow_control_chars: bool_value(input.allow_control_chars.unwrap_or(false)),
            origin_pane_title: input.origin_pane_title,
            origin_process_name: input.origin_process_name,
            status: "idle".to_owned(),
            last_error: None,
            max_steps_per_turn: input.max_steps_per_turn.unwrap_or(25),
            created_at: now.clone(),
            updated_at: now,
        };
        agent_sessions::Entity::insert(agent_sessions::ActiveModel {
            id: Set(model.id.clone()),
            title: Set(model.title.clone()),
            device_id: Set(model.device_id.clone()),
            pane_id: Set(model.pane_id.clone()),
            provider_id: Set(model.provider_id.clone()),
            model_id: Set(model.model_id.clone()),
            system_prompt: Set(model.system_prompt.clone()),
            write_mode: Set(model.write_mode.clone()),
            use_provider_web_search: Set(model.use_provider_web_search),
            provider_hosted_tools: Set(model.provider_hosted_tools.clone()),
            allow_control_chars: Set(model.allow_control_chars),
            origin_pane_title: Set(model.origin_pane_title.clone()),
            origin_process_name: Set(model.origin_process_name.clone()),
            status: Set(model.status.clone()),
            last_error: Set(model.last_error.clone()),
            max_steps_per_turn: Set(model.max_steps_per_turn),
            created_at: Set(model.created_at.clone()),
            updated_at: Set(model.updated_at.clone()),
        })
        .exec_without_returning(self.database.orm())
        .await?;
        self.get_agent_session_by_id(&model.id)
            .await?
            .ok_or(RepositoryError::MissingAfterWrite("agent session"))
    }

    pub async fn get_agent_session_by_id(
        &self,
        id: &str,
    ) -> RepositoryResult<Option<agent_sessions::Model>> {
        Ok(agent_sessions::Entity::find_by_id(id.to_owned())
            .one(self.database.orm())
            .await?)
    }

    pub async fn get_all_agent_sessions(&self) -> RepositoryResult<Vec<agent_sessions::Model>> {
        Ok(agent_sessions::Entity::find()
            .order_by_desc(agent_sessions::Column::UpdatedAt)
            .all(self.database.orm())
            .await?)
    }

    pub async fn get_agent_sessions_by_status(
        &self,
        status: &str,
    ) -> RepositoryResult<Vec<agent_sessions::Model>> {
        Ok(agent_sessions::Entity::find()
            .filter(agent_sessions::Column::Status.eq(status))
            .order_by_desc(agent_sessions::Column::UpdatedAt)
            .all(self.database.orm())
            .await?)
    }

    pub async fn update_agent_session(
        &self,
        id: &str,
        updates: AgentSessionUpdate,
    ) -> RepositoryResult<Option<agent_sessions::Model>> {
        let mut model = agent_sessions::ActiveModel {
            id: Set(id.to_owned()),
            updated_at: Set(now_iso()),
            ..Default::default()
        };
        if let Some(value) = updates.title {
            model.title = Set(value);
        }
        if let Some(value) = updates.device_id {
            model.device_id = Set(value);
        }
        if let Some(value) = updates.pane_id {
            model.pane_id = Set(value);
        }
        if let Some(value) = updates.provider_id {
            model.provider_id = Set(value);
        }
        if let Some(value) = updates.model_id {
            model.model_id = Set(value);
        }
        if let Some(value) = updates.system_prompt {
            model.system_prompt = Set(value);
        }
        if let Some(value) = updates.write_mode {
            model.write_mode = Set(value);
        }
        if let Some(value) = updates.use_provider_web_search {
            model.use_provider_web_search = Set(bool_value(value));
        }
        if let Some(value) = updates.provider_hosted_tools {
            model.provider_hosted_tools = Set(json_string_list(&value));
        }
        if let Some(value) = updates.allow_control_chars {
            model.allow_control_chars = Set(bool_value(value));
        }
        if let Some(value) = updates.status {
            model.status = Set(value);
        }
        if let Some(value) = updates.last_error {
            model.last_error = Set(value);
        }
        if let Some(value) = updates.max_steps_per_turn {
            model.max_steps_per_turn = Set(value);
        }
        agent_sessions::Entity::update_many()
            .set(model)
            .filter(agent_sessions::Column::Id.eq(id))
            .exec(self.database.orm())
            .await?;
        self.get_agent_session_by_id(id).await
    }

    pub async fn delete_agent_session(&self, id: &str) -> RepositoryResult<()> {
        agent_sessions::Entity::delete_by_id(id.to_owned())
            .exec(self.database.orm())
            .await?;
        Ok(())
    }

    pub async fn append_agent_message(
        &self,
        session_id: &str,
        role: &str,
        content: Json,
    ) -> RepositoryResult<agent_messages::Model> {
        let transaction = self.database.begin().await?;
        let id = new_id();
        let created_at = now_iso();
        let result = async {
            let max_seq = agent_messages::Entity::find()
                .filter(agent_messages::Column::SessionId.eq(session_id))
                .select_only()
                .column_as(agent_messages::Column::Seq.max(), "max_seq")
                .into_tuple::<Option<i64>>()
                .one(transaction.orm())
                .await?
                .flatten()
                .unwrap_or(-1);
            agent_messages::Entity::insert(agent_messages::ActiveModel {
                id: Set(id.clone()),
                session_id: Set(session_id.to_owned()),
                seq: Set(max_seq + 1),
                role: Set(role.to_owned()),
                content: Set(content.to_string()),
                created_at: Set(created_at.clone()),
            })
            .exec_without_returning(transaction.orm())
            .await?;
            RepositoryResult::Ok(())
        }
        .await;
        if let Err(error) = result {
            return rollback(transaction, error).await;
        }
        transaction.commit().await?;
        self.get_agent_message_by_id(&id)
            .await?
            .ok_or(RepositoryError::MissingAfterWrite("agent message"))
    }

    async fn get_agent_message_by_id(
        &self,
        id: &str,
    ) -> RepositoryResult<Option<agent_messages::Model>> {
        Ok(agent_messages::Entity::find_by_id(id.to_owned())
            .one(self.database.orm())
            .await?)
    }

    pub async fn list_agent_messages(
        &self,
        session_id: &str,
        after_seq: Option<i64>,
    ) -> RepositoryResult<Vec<agent_messages::Model>> {
        let mut query =
            agent_messages::Entity::find().filter(agent_messages::Column::SessionId.eq(session_id));
        if let Some(after_seq) = after_seq {
            query = query.filter(agent_messages::Column::Seq.gt(after_seq));
        }
        Ok(query
            .order_by_asc(agent_messages::Column::Seq)
            .all(self.database.orm())
            .await?)
    }

    pub async fn get_max_agent_message_seq(&self, session_id: &str) -> RepositoryResult<i64> {
        Ok(agent_messages::Entity::find()
            .filter(agent_messages::Column::SessionId.eq(session_id))
            .select_only()
            .column_as(agent_messages::Column::Seq.max(), "max_seq")
            .into_tuple::<Option<i64>>()
            .one(self.database.orm())
            .await?
            .flatten()
            .unwrap_or(-1))
    }

    pub async fn enqueue_agent_message(
        &self,
        session_id: &str,
        text: &str,
    ) -> RepositoryResult<agent_queued_messages::Model> {
        let transaction = self.database.begin().await?;
        let id = new_id();
        let created_at = now_iso();
        let result = async {
            let max_seq = agent_queued_messages::Entity::find()
                .filter(agent_queued_messages::Column::SessionId.eq(session_id))
                .select_only()
                .column_as(agent_queued_messages::Column::Seq.max(), "max_seq")
                .into_tuple::<Option<i64>>()
                .one(transaction.orm())
                .await?
                .flatten()
                .unwrap_or(-1);
            agent_queued_messages::Entity::insert(agent_queued_messages::ActiveModel {
                id: Set(id.clone()),
                session_id: Set(session_id.to_owned()),
                seq: Set(max_seq + 1),
                text: Set(text.to_owned()),
                created_at: Set(created_at.clone()),
            })
            .exec_without_returning(transaction.orm())
            .await?;
            RepositoryResult::Ok(())
        }
        .await;
        if let Err(error) = result {
            return rollback(transaction, error).await;
        }
        transaction.commit().await?;
        self.get_queued_agent_message_by_id(&id)
            .await?
            .ok_or(RepositoryError::MissingAfterWrite("queued agent message"))
    }

    pub async fn list_queued_agent_messages(
        &self,
        session_id: &str,
    ) -> RepositoryResult<Vec<agent_queued_messages::Model>> {
        Ok(agent_queued_messages::Entity::find()
            .filter(agent_queued_messages::Column::SessionId.eq(session_id))
            .order_by_asc(agent_queued_messages::Column::Seq)
            .all(self.database.orm())
            .await?)
    }

    pub async fn get_queued_agent_message_by_id(
        &self,
        id: &str,
    ) -> RepositoryResult<Option<agent_queued_messages::Model>> {
        Ok(agent_queued_messages::Entity::find_by_id(id.to_owned())
            .one(self.database.orm())
            .await?)
    }

    pub async fn update_queued_agent_message(
        &self,
        id: &str,
        text: &str,
    ) -> RepositoryResult<Option<agent_queued_messages::Model>> {
        agent_queued_messages::Entity::update_many()
            .set(agent_queued_messages::ActiveModel {
                text: Set(text.to_owned()),
                ..Default::default()
            })
            .filter(agent_queued_messages::Column::Id.eq(id))
            .exec(self.database.orm())
            .await?;
        self.get_queued_agent_message_by_id(id).await
    }

    pub async fn delete_queued_agent_message(&self, id: &str) -> RepositoryResult<()> {
        agent_queued_messages::Entity::delete_by_id(id.to_owned())
            .exec(self.database.orm())
            .await?;
        Ok(())
    }

    pub async fn delete_all_queued_agent_messages(&self, session_id: &str) -> RepositoryResult<()> {
        agent_queued_messages::Entity::delete_many()
            .filter(agent_queued_messages::Column::SessionId.eq(session_id))
            .exec(self.database.orm())
            .await?;
        Ok(())
    }

    pub async fn drain_queued_agent_messages(
        &self,
        session_id: &str,
    ) -> RepositoryResult<Vec<agent_messages::Model>> {
        let transaction = self.database.begin().await?;
        let result = async {
            let queued = agent_queued_messages::Entity::find()
                .filter(agent_queued_messages::Column::SessionId.eq(session_id))
                .order_by_asc(agent_queued_messages::Column::Seq)
                .all(transaction.orm())
                .await?;
            if queued.is_empty() {
                return RepositoryResult::Ok(Vec::new());
            }

            let max_seq = agent_messages::Entity::find()
                .filter(agent_messages::Column::SessionId.eq(session_id))
                .select_only()
                .column_as(agent_messages::Column::Seq.max(), "max_seq")
                .into_tuple::<Option<i64>>()
                .one(transaction.orm())
                .await?
                .flatten()
                .unwrap_or(-1);
            let mut messages = Vec::with_capacity(queued.len());
            for (offset, item) in queued.into_iter().enumerate() {
                let message = agent_messages::Model {
                    id: new_id(),
                    session_id: session_id.to_owned(),
                    seq: max_seq + offset as i64 + 1,
                    role: "user".to_owned(),
                    content: json!({"role":"user","content":item.text}).to_string(),
                    created_at: now_iso(),
                };
                agent_messages::Entity::insert(agent_messages::ActiveModel {
                    id: Set(message.id.clone()),
                    session_id: Set(message.session_id.clone()),
                    seq: Set(message.seq),
                    role: Set(message.role.clone()),
                    content: Set(message.content.clone()),
                    created_at: Set(message.created_at.clone()),
                })
                .exec_without_returning(transaction.orm())
                .await?;
                messages.push(message);
            }
            agent_queued_messages::Entity::delete_many()
                .filter(agent_queued_messages::Column::SessionId.eq(session_id))
                .exec(transaction.orm())
                .await?;
            RepositoryResult::Ok(messages)
        }
        .await;
        let messages = match result {
            Ok(messages) => messages,
            Err(error) => return rollback(transaction, error).await,
        };
        transaction.commit().await?;
        Ok(messages)
    }

    pub async fn create_agent_confirmation(
        &self,
        input: CreateAgentConfirmationInput,
    ) -> RepositoryResult<agent_confirmations::Model> {
        let model = agent_confirmations::Model {
            id: input.id.unwrap_or_else(new_id),
            session_id: input.session_id,
            tool_name: input.tool_name,
            tool_call_id: input.tool_call_id,
            input_json: input.input_json.to_string(),
            status: "pending".to_owned(),
            reason: None,
            decided_at: None,
            created_at: now_iso(),
        };
        agent_confirmations::Entity::insert(agent_confirmations::ActiveModel {
            id: Set(model.id.clone()),
            session_id: Set(model.session_id.clone()),
            tool_name: Set(model.tool_name.clone()),
            tool_call_id: Set(model.tool_call_id.clone()),
            input_json: Set(model.input_json.clone()),
            status: Set(model.status.clone()),
            reason: Set(model.reason.clone()),
            decided_at: Set(model.decided_at.clone()),
            created_at: Set(model.created_at.clone()),
        })
        .exec_without_returning(self.database.orm())
        .await?;
        self.get_agent_confirmation_by_id(&model.id)
            .await?
            .ok_or(RepositoryError::MissingAfterWrite("agent confirmation"))
    }

    pub async fn get_agent_confirmation_by_id(
        &self,
        id: &str,
    ) -> RepositoryResult<Option<agent_confirmations::Model>> {
        Ok(agent_confirmations::Entity::find_by_id(id.to_owned())
            .one(self.database.orm())
            .await?)
    }

    pub async fn list_pending_agent_confirmations(
        &self,
        session_id: &str,
    ) -> RepositoryResult<Vec<agent_confirmations::Model>> {
        Ok(agent_confirmations::Entity::find()
            .filter(agent_confirmations::Column::SessionId.eq(session_id))
            .filter(agent_confirmations::Column::Status.eq("pending"))
            .order_by_asc(agent_confirmations::Column::CreatedAt)
            .all(self.database.orm())
            .await?)
    }

    pub async fn decide_agent_confirmation(
        &self,
        id: &str,
        decision: AgentConfirmationDecision,
    ) -> RepositoryResult<Option<agent_confirmations::Model>> {
        let result = agent_confirmations::Entity::update_many()
            .set(agent_confirmations::ActiveModel {
                status: Set(decision.status),
                reason: Set(decision.reason),
                decided_at: Set(Some(now_iso())),
                ..Default::default()
            })
            .filter(agent_confirmations::Column::Id.eq(id))
            .filter(agent_confirmations::Column::Status.eq("pending"))
            .exec(self.database.orm())
            .await?;
        if result.rows_affected == 0 {
            return Ok(None);
        }
        self.get_agent_confirmation_by_id(id).await
    }
}
